File: app/src/main/java/com/mrboomdev/awery/ui/ThemeManager.java
package com.mrboomdev.awery.ui;

import static android.content.Context.UI_MODE_SERVICE;

import static com.mrboomdev.awery.data.settings.NicePreferences.getPrefs;

import android.app.Activity;
import android.app.UiModeManager;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;

import com.google.android.material.color.DynamicColors;
import com.google.android.material.color.DynamicColorsOptions;
import com.mrboomdev.awery.R;
import com.mrboomdev.awery.generated.AwerySettings;

import java.util.Objects;

public class ThemeManager {

	public static boolean isMaterialYou() {
		if(!AwerySettings.USE_MATERIAL_YOU.exists()) {
			boolean isMaterialYouSupported = DynamicColors.isDynamicColorAvailable();

			getPrefs().setValue(AwerySettings.USE_MATERIAL_YOU, isMaterialYouSupported).saveAsync();
			return isMaterialYouSupported;
		}

		return AwerySettings.USE_MATERIAL_YOU.getValue();
	}

	public static void applyApp(Context context) {
		var isDarkModeEnabled = AwerySettings.USE_DARK_THEME.getValue(null);

		if(isDarkModeEnabled != null) {
			if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
				((UiModeManager) context.getSystemService(UI_MODE_SERVICE))
						.setApplicationNightMode(isDarkModeEnabled
								? UiModeManager.MODE_NIGHT_YES : UiModeManager.MODE_NIGHT_NO);
			} else {
				AppCompatDelegate.setDefaultNightMode(isDarkModeEnabled
						? AppCompatDelegate.MODE_NIGHT_YES
						: AppCompatDelegate.MODE_NIGHT_NO);
			}
		}
	}

	public static void apply(Activity activity, Bitmap bitmap) {
		boolean useOLED = AwerySettings.USE_AMOLED_THEME.getValue();
		boolean useColorsFromPoster = AwerySettings.EXTRACT_BANNER_COLOR.getValue();
		boolean useMaterialYou = isMaterialYou();

		if(useMaterialYou || (useColorsFromPoster && bitmap != null)) {
			applyMaterialYou(activity, bitmap, useOLED);
			return;
		}

		var savedTheme = Objects.requireNonNullElse(
				AwerySettings.THEME_COLOR_PALETTE.getValue(),
				AwerySettings.ThemeColorPalette_Values.RED);

		// In light mode there is no amoled theme so we need to check the current night mode
		if(useOLED && (activity.getResources().getConfiguration().uiMode
				& Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_NO) {
			activity.setTheme(getThemeRes(savedTheme, false));
			return;
		}

		activity.setTheme(getThemeRes(savedTheme, useOLED));
	}

	public static void apply(Activity activity) {
		apply(activity, null);
	}

	private static void applyMaterialYou(Activity activity, Bitmap bitmap, boolean useOLED) {
		var options = new DynamicColorsOptions.Builder();

		if(bitmap != null) {
			options.setContentBasedSource(bitmap);
		}

		if(useOLED) {
			options.setThemeOverlay(R.style.AmoledThemeOverlay);
		}

		DynamicColors.applyToActivityIfAvailable(activity, options.build());
	}

	private static int getThemeRes(@NonNull AwerySettings.ThemeColorPalette_Values theme, boolean isAmoled) {
		return switch(theme) {
			case RED -> isAmoled ? R.style.Theme_Awery_RedOLED : R.style.Theme_Awery_Red;
			case PINK -> isAmoled ? R.style.Theme_Awery_PinkOLED : R.style.Theme_Awery_Pink;
			case PURPLE -> isAmoled ? R.style.Theme_Awery_PurpleOLED : R.style.Theme_Awery_Purple;
			case BLUE -> isAmoled ? R.style.Theme_Awery_BlueOLED : R.style.Theme_Awery_Blue;
			case GREEN -> isAmoled ? R.style.Theme_Awery_GreenOLED : R.style.Theme_Awery_Green;
			case MONOCHROME -> isAmoled ? R.style.Theme_Awery_MonochromeOLED : R.style.Theme_Awery_Monochrome;
		};
	}
}

File: androidApp/src/main/kotlin/com/mrboomdev/awery/app/theme/ThemeManager.kt
package com.mrboomdev.awery.app.theme

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.app.UiModeManager
import android.content.res.Configuration
import android.os.Build
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatDelegate
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import com.google.android.material.color.DynamicColors
import com.google.android.material.color.DynamicColorsOptions
import com.mrboomdev.awery.R
import com.mrboomdev.awery.app.App.Companion.isTv
import com.mrboomdev.awery.app.AweryLifecycle.Companion.anyContext
import com.mrboomdev.awery.generated.AwerySettings

object ThemeManager {
	val currentColorPalette: AwerySettings.ThemeColorPaletteValue
		get() = AwerySettings.THEME_COLOR_PALETTE.value ?: resetPalette()

	val isDarkModeEnabled: Boolean
		get() {
			val config = anyContext.resources.configuration
			return (config.uiMode and Configuration.UI_MODE_NIGHT_YES) == Configuration.UI_MODE_NIGHT_YES
		}

	fun Application.applyTheme() {
		var isDarkModeEnabled = AwerySettings.USE_DARK_THEME.value

		// Light theme on tv is really a bad thing.
		if(isTv) isDarkModeEnabled = true

		if(isDarkModeEnabled != null) {
			if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
				getSystemService(UiModeManager::class.java)
					.setApplicationNightMode(
						if(isDarkModeEnabled) UiModeManager.MODE_NIGHT_YES
						else UiModeManager.MODE_NIGHT_NO
					)
			} else {
				AppCompatDelegate.setDefaultNightMode(
					if(isDarkModeEnabled) AppCompatDelegate.MODE_NIGHT_YES
					else AppCompatDelegate.MODE_NIGHT_NO
				)
			}
		}
	}

	fun Activity.applyTheme() {
		val isAmoled = AwerySettings.USE_AMOLED_THEME.value
		val palette = currentColorPalette

		if(palette == AwerySettings.ThemeColorPaletteValue.MATERIAL_YOU) {
			return DynamicColors.applyToActivityIfAvailable(this,
				DynamicColorsOptions.Builder().apply {
					if(isAmoled) {
						setThemeOverlay(R.style.AmoledThemeOverlay)
					}
				}.build())
		}

		setTheme(getThemeRes(palette, isAmoled && isDarkModeEnabled))
	}

	/**
	 * Set an composable component as activity's content and apply an current theme.
	 */
	fun ComponentActivity.setThemedContent(content: @Composable () -> Unit) {
		setContent {
			var isDarkMode by remember { mutableStateOf(isDarkModeEnabled) }
			var isAmoledEnabled by remember { mutableStateOf(AwerySettings.USE_AMOLED_THEME.value) }
			
			val theme by remember { derivedStateOf { 
				object : AweryTheme {
					override var isDark: Boolean
						get() = isDarkMode
						set(value) { isDarkMode = value }
					
					override var isAmoled: Boolean
						get() = isDarkMode && isAmoledEnabled
						set(value) { isAmoledEnabled = value }
					
				}
			}}
			
			CompositionLocalProvider(LocalAweryTheme provides theme) {
				MobileTheme(
					palette = currentColorPalette,
					isDark = theme.isDark,
					isAmoled = theme.isAmoled
				) {
					TvTheme(
						palette = currentColorPalette,
						isDark = theme.isDark,
						isAmoled = theme.isAmoled
					) {
						Surface {
							content()
						}
					}
				}
			}
		}
	}

	@SuppressLint("PrivateResource")
	fun getThemeRes(theme: AwerySettings.ThemeColorPaletteValue, isAmoled: Boolean): Int {
		// Amoled theme breaks some colors in a light theme.
		val isReallyAmoled = isAmoled && isDarkModeEnabled

		return when(theme) {
			AwerySettings.ThemeColorPaletteValue.RED -> 
				if(isReallyAmoled) R.style.Theme_Awery_Red_Amoled else R.style.Theme_Awery_Red
			
			AwerySettings.ThemeColorPaletteValue.PINK -> 
				if(isReallyAmoled) R.style.Theme_Awery_Pink_Amoled else R.style.Theme_Awery_Pink
			
			AwerySettings.ThemeColorPaletteValue.PURPLE -> 
				if(isReallyAmoled) R.style.Theme_Awery_Purple_Amoled else R.style.Theme_Awery_Purple
			
			AwerySettings.ThemeColorPaletteValue.BLUE -> 
				if(isReallyAmoled) R.style.Theme_Awery_Blue_Amoled else R.style.Theme_Awery_Blue
			
			AwerySettings.ThemeColorPaletteValue.GREEN -> 
				if(isReallyAmoled) R.style.Theme_Awery_Green_Amoled else R.style.Theme_Awery_Green
			
			AwerySettings.ThemeColorPaletteValue.MONOCHROME -> 
				if(isReallyAmoled) R.style.Theme_Awery_Monochrome_Amoled else R.style.Theme_Awery_Monochrome
			
			AwerySettings.ThemeColorPaletteValue.MATERIAL_YOU -> 
				com.google.android.material.R.style.Theme_Material3_DynamicColors_DayNight
		}
	}

	private fun resetPalette(): AwerySettings.ThemeColorPaletteValue {
		val isMaterialYouSupported = DynamicColors.isDynamicColorAvailable()

		val theme = if(isMaterialYouSupported) {
			AwerySettings.ThemeColorPaletteValue.MATERIAL_YOU
		} else {
			AwerySettings.ThemeColorPaletteValue.RED
		}

		AwerySettings.THEME_COLOR_PALETTE.value = theme
		return theme
	}
}